network/p2p/throttler.go

```diff
@@ -0,0 +1,48 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package p2p
+
+import (
+	"context"
+	"sync"
+
+	"golang.org/x/time/rate"
+
+	"github.com/ava-labs/avalanchego/ids"
+)
+
+var _ Throttler = (*TokenBucketThrottler)(nil)
+
+type Throttler interface {
+	// Throttle blocks until there is [tokens] capacity for [nodeID]
+	Throttle(ctx context.Context, nodeID ids.NodeID, tokens int) error
+}
+
+func NewTokenBucketThrottler(refill rate.Limit, burst int) *TokenBucketThrottler {
+	return &TokenBucketThrottler{
+		refill:       refill,
+		burst:        burst,
+		tokenBuckets: make(map[ids.NodeID]*rate.Limiter),
+	}
+}
+
+type TokenBucketThrottler struct {
+	refill rate.Limit
+	burst  int
+
+	lock         sync.Mutex
+	tokenBuckets map[ids.NodeID]*rate.Limiter
```

Comment:
Are we concerned about the fact that we never remove stuff from this map? Maybe we could remove an entry if it's unused for a certain amount of time?

Reply:
Yeah it's wrong. I think it might make sense to manage this via `Connect` and `Disconnect` handlers.